File: ru/codes/java/chapter_greedy/fractional_knapsack.java
/**
 * File: fractional_knapsack.java
 * Created Time: 2023-07-20
 * Author: krahets (krahets@163.com)
 */

package chapter_greedy;

import java.util.Arrays;
import java.util.Comparator;

/* Предмет */
class Item {
    int w; // Вес предмета
    int v; // Стоимость предмета

    public Item(int w, int v) {
        this.w = w;
        this.v = v;
    }
}

public class fractional_knapsack {
    /* Дробный рюкзак: жадный алгоритм */
    static double fractionalKnapsack(int[] wgt, int[] val, int cap) {
        // Создать список предметов с двумя свойствами: вес и стоимость
        Item[] items = new Item[wgt.length];
        for (int i = 0; i < wgt.length; i++) {
            items[i] = new Item(wgt[i], val[i]);
        }
        // Отсортировать по удельной стоимости item.v / item.w в порядке убывания
        Arrays.sort(items, Comparator.comparingDouble(item -> -((double) item.v / item.w)));
        // Циклический жадный выбор
        double res = 0;
        for (Item item : items) {
            if (item.w <= cap) {
                // Если оставшейся вместимости достаточно, положить в рюкзак текущий предмет целиком
                res += item.v;
                cap -= item.w;
            } else {
                // Если оставшейся вместимости недостаточно, положить в рюкзак часть текущего предмета
                res += (double) item.v / item.w * cap;
                // Свободной вместимости больше не осталось, поэтому выйти из цикла
                break;
            }
        }
        return res;
    }

    public static void main(String[] args) {
        int[] wgt = { 10, 20, 30, 40, 50 };
        int[] val = { 50, 120, 150, 210, 240 };
        int cap = 50;

        // Жадный алгоритм
        double res = fractionalKnapsack(wgt, val, cap);
        System.out.println("Максимальная стоимость предметов без превышения вместимости рюкзака = " + res);
    }
}

File: ru/codes/go/chapter_greedy/fractional_knapsack.go
// File: fractional_knapsack.go
// Created Time: 2023-07-23
// Author: Reanon (793584285@qq.com)

package chapter_greedy

import "sort"

/* Предмет */
type Item struct {
	w int // Вес предмета
	v int // Стоимость предмета
}

/* Дробный рюкзак: жадный алгоритм */
func fractionalKnapsack(wgt []int, val []int, cap int) float64 {
	// Создать список предметов с двумя свойствами: вес и стоимость
	items := make([]Item, len(wgt))
	for i := 0; i < len(wgt); i++ {
		items[i] = Item{wgt[i], val[i]}
	}
	// Отсортировать по удельной стоимости item.v / item.w в порядке убывания
	sort.Slice(items, func(i, j int) bool {
		return float64(items[i].v)/float64(items[i].w) > float64(items[j].v)/float64(items[j].w)
	})
	// Циклический жадный выбор
	res := 0.0
	for _, item := range items {
		if item.w <= cap {
			// Если оставшейся вместимости достаточно, положить в рюкзак текущий предмет целиком
			res += float64(item.v)
			cap -= item.w
		} else {
			// Если оставшейся вместимости недостаточно, положить в рюкзак часть текущего предмета
			res += float64(item.v) / float64(item.w) * float64(cap)
			// Свободной вместимости больше не осталось, поэтому выйти из цикла
			break
		}
	}
	return res
}
